_drafts/installing-gdb.md

```diff
@@ -0,0 +1,514 @@
+embedded device, connected to some debug probe). That setup might look like
+this:
+
+[![img/installing-gdb/gdb_cross_target.drawio.svg](/img/installing-gdb/gdb_cross_target.drawio.svg)](/img/installing-gdb/gdb_cross_target.drawio.svg)
+
+We need a version of GDB that supports the target architecture of the program
+being debugged.
+
+## Summary of strategies
+
+I'm going to go through the various ways to install GDB with ARM support, but
+first here's a table summarizing the approaches:
+
+| Strategy                   | Binary or source install? | Pinned to version | Python support        |
+| -------------------------- | ------------------------- | ----------------- | --------------------- |
+| Build from source          | 📁source                  | ✅ yes            | ✅ (configurable)     |
+| Binaries from ARM          | 📦binary                  | ✅ yes            | ⚠️ varies per version |
+| System package manager     | 📦binary                  | ⚠️ varies         | ⚠️ varies             |
+| Conda                      | 📦binary                  | ✅ yes            | ✅ yes                |
+| Docker                     | 📦binary                  | ✅ yes            | ✅ yes                |
+| SDK-specific tools         | 📦binary                  | ⚠️ varies         | ✅ (usually yes)      |
+| xPack GNU Arm Embedded GCC | 📦binary                  | ✅ yes            | ✅ yes                |
+| crosstool-NG               | 📁source                  | ✅ yes            | ✅ (configurable)     |
+
+## Details on each strategy
+
+### Build from source
+
+[Building a custom copy of GDB]({% post_url 2020-04-08-gnu-binutils %}) is one
+way to enable support for the necessary architecture.
+
+You might for example build GDB enabling all targets:
+
+```bash
+❯ ./configure
+  --target="arm-elf-linux"
+  --enable-targets=all
+  --with-python
+❯ make -j $(nproc)
+```
+
+This results in a nicely tuned copy of GDB- you can select only the features
+needed, enable different optimizations, select the correct Python version to
+integrate, etc., and are in full control of your software supply chain!
+
+The downside for from-source builds (in general):
+
+- can be challenging to build for Windows or other non-Unix platforms
+- need to maintain build tooling + binary repositories
+- using custom builds, you might encounter edge cases or warts that have been
+  solved in more widely used copies of the same software
+
+> Quick note- if you want to see which architectures a given `gdb` binary
+> supports, here's a shell "one-liner" that can be useful (taken from
+> [here](https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=gdb/gdb_buildall.sh;hb=4a94e36819485cdbd50438f800d1e478156a4889#l206)):
+>
+> ```bash
+> gdb --batch -nx --ex 'set architecture' 2>&1 | \
+>   tail -n 1 | \
+>   sed 's/auto./\n/g' | \
+>   sed 's/,/\n/g' | \
+>   sed 's/Requires an argument. Valid arguments are/\n/g' | \
+>   sed '/^[ ]*$/d' | sort | uniq
```

Comment:
I don't have a better one line for you but note this only works on linux -- OG sed doesn't support emitting \n :/

Reply:
Urk 😂 always gets me! I'll make a note